wp-includes/js/service-worker-navigation-routing.js

```diff
@@ -1,181 +1,249 @@
-						let init = {
-							status: response.status,
-							statusText: response.statusText,
-							headers: response.headers
-						};
+				return response.text().then( function( text ) {
+					let init = {
+						status: response.status,
+						statusText: response.statusText,
+						headers: response.headers
+					};
 
-						const body = text.replace( /<!--WP_SERVICE_WORKER_ERROR_MESSAGE-->/, errorMessages.default );
+					const body = text.replace( /<!--WP_SERVICE_WORKER_ERROR_MESSAGE-->/, errorMessages.default );
 
-						return new Response( body, init );
-					} );
+					return new Response( body, init );
 				} );
+			} );
+		};
+
+		/*
+		 * If navigation preload is enabled, use the preload request instead of doing another fetch.
+		 * This prevents requests from being duplicated. See <https://github.com/xwp/pwa-wp/issues/67>.
+		 */
+		if ( event.preloadResponse ) {
+			try {
+				const response = await event.preloadResponse;
+				if ( response ) {
+					responsePreloaded = true;
+					return handleResponse( response );
+				}
+			} catch ( error ) {
+				responsePreloaded = true;
+				return sendOfflineResponse();
+			}
+		}
+
+		const navigationCacheStrategy = wp.serviceWorker.strategies[ CACHING_STRATEGY ]( CACHING_STRATEGY_ARGS );
+
+		if ( canStreamResponse() ) {
+			const streamHeaderFragmentURL = STREAM_HEADER_FRAGMENT_URL;
+			const precacheStrategy = wp.serviceWorker.strategies.cacheFirst({
+				cacheName: wp.serviceWorker.core.cacheNames.precache,
+			});
+
+			const url = new URL( event.request.url );
+			url.searchParams.append( STREAM_HEADER_FRAGMENT_QUERY_VAR, 'body' );
+			const init = {
+				mode: 'same-origin'
 			};
+			const copiedProps = [
+				'method',
+				'headers',
+				'credentials',
+				'cache',
+				'redirect',
+				'referrer',
+				'integrity',
+			];
+			for ( const initProp of copiedProps ) {
+				init[ initProp ] = event.request[ initProp ];
+			}
+			const request = new Request( url.toString(), init );
+			const stream = wp.serviceWorker.streams.concatenateToResponse([
+				precacheStrategy.makeRequest({ request: streamHeaderFragmentURL }),
+				navigationCacheStrategy.makeRequest( { request } )
+					.then( handleResponse )
+					.catch( sendOfflineResponse ),
+			]);
+
+			return stream.response;
+		} else {
+			return navigationCacheStrategy.makeRequest( { request: event.request } )
+				.then( handleResponse )
+				.catch( sendOfflineResponse );
+		}
+	}
 
-			/*
-			 * If navigation preload is enabled, use the preload request instead of doing another fetch.
-			 * This prevents requests from being duplicated. See <https://github.com/xwp/pwa-wp/issues/67>.
+	const blacklist = NAVIGATION_BLACKLIST_PATTERNS.map( ( pattern ) => new RegExp( pattern ) );
+	if ( navigationRouteEntry ) {
+		wp.serviceWorker.routing.registerNavigationRoute(
+			navigationRouteEntry.url,
+			{ blacklist }
+		);
+
+		class FetchNavigationRoute extends wp.serviceWorker.routing.Route {
```

Comment:
It would be nice if Workbox had a Whitelist/Blacklist route that this and `NavigationRoute` could extend. /cc @jeffposnick